snuba/state/cache/abstract.py

```diff
@@ -0,0 +1,60 @@
+from abc import ABC, abstractmethod
+from typing import Callable, Generic, Optional, TypeVar
+
+
+T = TypeVar("T")
+
+
+class ExecutionError(Exception):
+    pass
+
+
+class ExecutionTimeoutError(ExecutionError):
+    pass
+
+
+class Cache(Generic[T], ABC):
+    @abstractmethod
+    def get(self, key: str) -> Optional[T]:
+        """
+        Gets a value from the cache.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def set(self, key: str, value: T) -> None:
+        """
+        Sets a value in the cache.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_or_execute(self, key: str, function: Callable[[], T], timeout: int) -> T:
```

Comment:
> So I would still change the name to anything that includes `readthrough` because that would immediately tell people a lot what to expect from this method.

Alright, I will update to reflect that.

> Also the `execute` part of `get_or_execute` seems very contextual to executing a query, while this could just be fetching data from somewhere.

I was thinking of it more as "execute the function", but I guess that didn't really come through the way that I thought it would.

> About splitting the classes. Honestly I would have never thought the driver for keeping them together was to allow the results to be compatible with get_or_execute. The fact this method is not yet used in the code base does not help to understand why we need the old get/set methods as well.

Yeah, I can see how this would be confusing out of context.

> My first reaction would be that we should not be accessing a read through cache through raw get/set methods (we could set a value with `set` that gets overwritten by a stale value we are waiting for the results of via get_or_execute). Though if there are use cases to expose get/set to access the same keys we access with the read through method, I am ok with that.

I generally agree that the two shouldn't be intermixed, at least on the same key.

I pictured this as just a generic cache implementation, but in reality it's only used for the query cache, so if we're confident that a runtime switch isn't required for enabling/disabling the deduplication, I can remove `get` and `set` in the change that actually integrates this function into the query path. (I'm not sure that I am that confident yet, but your opinion might be different.) Otherwise, I can try to clarify somewhere that the two access patterns should be compatible, caveats aside as you noted. Let me know what you think makes sense.

Reply:
Changed the name and slightly updated the documentation with c88bcba and f134bbb.

Leaving this unresolved because of open questions from the last paragraph of the last comment.